dynamiqs/time_tensor.py

```diff
@@ -239,3 +278,170 @@ def __add__(self, other: Tensor | TimeTensor) -> TimeTensor:
             return CallableTimeTensor(f, f0)
         else:
             return NotImplemented
+
+
+class PWCTimeTensor(TimeTensor):
+    # Arbitrary sum of PWC tensors.
+    #
+    # A single PWC tensor is defined by a tuple of 3 tensors (times, values, tensor),
+    # where
+    # - times: (nv+1) are the time points between which the PWC tensor take constant
+    #          values, where nv is the number of time intervals
+    # - values: (nv, ...) are the constant complex values for each time interval, where
+    #           (...) is an arbitrary batching size
+    # - tensor: (n, n) is the tensor value
+    #
+    # To support additions of PWC tensors with non-aligned `times` values we augment
+    # the `values` and `tensor` objects with one dimension `npwc`, which is the number
+    # of summed PWC tensors. Upon addition between two PWC tensors, we merge the
+    # `times` and `values` tensors appropriately, and concatenate the `tensor` objects.
+    # In addition, we keep track of a `static` tensor for addition with a constant
+    # tensor.
+
+    def __init__(
+        self,
+        times: Tensor,
+        values: Tensor,
+        tensors: Tensor,
+        static: Tensor | None = None,  # constant part
+    ):
+        # Dimensions:
+        # - nv: number of time intervals
+        # - ...: values batching (`self.values` carries all the shape transformation)
+        # - npwc: number of piecewise constant tensors
+        # - n: Hilbert space dimension
+
+        self.times = times  # (nv+1)
+
+        self.values = values  # (npwc, nv, ...)
+        self.npwc = values.shape[0]
+        self.nv = values.shape[1]
+        self.dots = values.shape[2:]  # (...)
```

Comment:
```suggestion
        self.nbatch = values.shape[2:]  # (...)
```
(maybe more explicit)